src/protocol/attribute.rs

```diff
@@ -0,0 +1,224 @@
+use super::{AttributionModel, RecordId};
+use crate::{
+    error::BoxError,
+    field::Field,
+    helpers::mesh::{Gateway, Mesh},
+    protocol::context::ProtocolContext,
+    secret_sharing::Replicated,
+};
+use futures::future::try_join_all;
+
+/// Obilivious attribution protocol
+///
+/// Obliviously attribute trigger events to corresponding source events for different breakdowns in a MPC.
+/// There are several common attribution models such as Last Touch, Even Credit and Exponential Decay.
+/// https://github.com/patcg-individual-drafts/ipa/blob/main/IPA-End-to-End.md#oblivious-last-touch-attribution
+#[allow(dead_code)]
+pub struct Attribute<'a, G> {
+    ctx: &'a ProtocolContext<'a, G, AttributionModel>,
+    step: AttributionModel,
+}
+
+impl<'a, G> Attribute<'a, G> {
+    const _IS_TRIGGER_BIT: usize = 2;
+    const _HELPER_BIT: usize = 3;
+    const _VALUE: usize = 5;
+
+    #[allow(dead_code)]
+    pub fn new(ctx: &'a ProtocolContext<'a, G, AttributionModel>, step: AttributionModel) -> Self {
+        Self { ctx, step }
+    }
+
+    /// Executes the protocol and produces an attribution result for the model specified by `step` in `Attribute::new`.
+    /// Attribution result will be used to calculate the aggregation in the next step.
+    #[allow(dead_code)]
+    pub async fn execute<M, F>(
+        &self,
+        input: &'a [Vec<Replicated<F>>],
+    ) -> Result<Vec<Replicated<F>>, BoxError>
+    where
+        M: Mesh,
+        G: Gateway<M, AttributionModel>,
+        F: Field,
+    {
+        match self.step {
+            AttributionModel::LastTouch => self.last_touch_attribution(input).await,
+        }
+    }
+
+    async fn last_touch_attribution<M, F>(
+        &self,
+        input: &'a [Vec<Replicated<F>>],
+    ) -> Result<Vec<Replicated<F>>, BoxError>
+    where
+        M: Mesh,
+        G: Gateway<M, AttributionModel>,
+        F: Field,
+    {
+        // TOOD: input length range check. Can we assume `assert!(num_rows < u32::MAX)`?
+        let num_rows = input[0].len();
+
+        // 1. Create Credit and Stop bit vectors
+        let mut credits: Vec<Replicated<F>> = Vec::with_capacity(num_rows);
+        input[Self::_VALUE].iter().for_each(|x| credits.push(*x));
+
+        let one = self.ctx.gateway.get_channel(self.step).share_of_one::<F>();
+        let mut stop_bits: Vec<Replicated<F>> = vec![one; num_rows];
+
+        // 2. Accumulate
+        //
+        // Last touch accumulation algorithm:
+        //     b = stop_bit * successor.helper_bit * successor.trigger_bit;
+        //     credit = credit + b * successor.credit;
+        //     stop_bit = b * successor.stop_bit;
+        let mut step_size = 1;
+        while step_size < num_rows {
```

Comment:
you can replace these three lines (init, while, *= 2) with a one liner that creates an iterator over powers of 2. 

```rust
        // generate powers of 2 that fit into input len. If num_rows is 15, this will produce [1, 2, 4, 8]
        for step_size in std::iter::successors(Some(1u32), |prev| prev.checked_mul(2)).take_while(|&v| v < num_rows) {
}
```